src/solve/solver.rs

```diff
@@ -136,7 +136,12 @@ impl Solver {
             // if all the components of the cycle also have coinductive semantics, we accept
             // the cycle `(?0: AutoTrait) :- ... :- (?0: AutoTrait)` as an infinite proof for
             // `?0: AutoTrait` and we do not perform any substitution.
-            if self.stack.iter().skip(index).all(|s| s.goal.is_coinductive(&*self.program)) {
+            if self.stack.iter()
+                         .skip(index)
+                         .map(|s| &s.goal)
+                         .chain(Some(&goal))
```

Comment:
Actually I realised that from a correctness point of view, that's the same thing because we already know that there is a cycle starting at index `index`, so the very first element of our iterator is also our current goal already. Anyway, let's keep it that way :)